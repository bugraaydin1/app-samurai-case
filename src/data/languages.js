export const languages = [
	{
		title: "English",
		code: "en",
		enterText: "Enter text",
		translation: "Translation",
		loading: "Translating...",
		noHistory: "History not found",
		history: "Translation History",
		flag: "ğŸ‡¬ğŸ‡§",
	},
	{
		title: "TÃ¼rkÃ§e",
		code: "tr",
		enterText: "Metin giriniz",
		translation: "Ã‡eviri",
		loading: "Ã‡eviri YapÄ±lÄ±yor...",
		noHistory: "GeÃ§miÅŸ arama bulunamadÄ±",
		history: "Ã‡eviri GeÃ§miÅŸi",
		flag: "ğŸ‡¹ğŸ‡·",
	},
];
