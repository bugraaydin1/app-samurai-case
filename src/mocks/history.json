{
	"history": [
		{ "en": "car", "tr": "araba" },
		{ "en": "home", "tr": "ev" }
	]
}
