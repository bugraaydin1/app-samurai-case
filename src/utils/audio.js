/**
 * Speaks supplied text
 * @param {string} text
 * @param {string} languageCode
 * @param {boolean=} loading
 */

export const speakTTS = (text, languageCode, loading) => {
	const voices = speechSynthesis?.getVoices();

	if (!text) return;

	if (window.speechSynthesis !== undefined) {
		const utterance = new SpeechSynthesisUtterance();

		utterance.volume = 0.5;
		utterance.rate = 1;
		utterance.pitch = 1;
		utterance.text = text;
		utterance.lang = languageCode;
		utterance.voice = voices.find((voice) => voice.lang.startsWith(languageCode));

		!loading && speechSynthesis.speak(utterance);
	} else {
		console.warn("Text To Speech is not supported ‚ùé");
	}
};
