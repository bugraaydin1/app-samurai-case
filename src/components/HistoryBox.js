import { useEffect, useState } from "react";
import { speakTTS } from "../utils/audio";
import Speaker from "./svg/Speaker";

const HistoryBox = ({ selectedLanguage }) => {
	const [history, setHistory] = useState([{ source: "car", target: "home" }]);

	const getStorageHistory = () => {
		return JSON.parse(localStorage.getItem(`${selectedLanguage.code}`) ?? "[]");
	};

	useEffect(() => {
		setHistory(getStorageHistory());
	}, [selectedLanguage]);

	return (
		<div className="history-wrapper">
			<h6 className="sticky-title">
				{selectedLanguage.flag} {selectedLanguage.history}
			</h6>

			{history.length ? (
				history.map((item, i) => (
					<div className="flex row" key={`${item.source}-${i}`}>
						<div
							className="translation-item icon-button"
							onClick={() => speakTTS(item.source, selectedLanguage.code)}
						>
							<Speaker size="16px" />
							{item.source} â†’
						</div>

						<div className="translation-item">{item.target}</div>
					</div>
				))
			) : (
				<div>{selectedLanguage.noHistory}</div>
			)}
		</div>
	);
};

export default HistoryBox;
