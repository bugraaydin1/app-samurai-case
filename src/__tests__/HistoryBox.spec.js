import * as React from "react";
import { screen, render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";

import HistoryBox from "../components/HistoryBox";
import { languages } from "../data/languages";
import { history } from "../mocks/history.json";
import * as audio from "../utils/audio";

describe("HistoryBox", () => {
	const language = languages[0];

	const historyComponent = (props) => (
		<HistoryBox loading={false} selectedLanguage={language} {...props} />
	);

	it("renders correctly", () => {
		const historyBox = renderer.create(historyComponent()).toJSON();

		expect(historyBox).toMatchSnapshot();
	});

	it("title is correct", () => {
		render(historyComponent());

		const input = screen.getByText(`${language.flag} ${language.history}`);

		expect(input).toBeInTheDocument();
	});

	it("when no history shows relevant text", () => {
		render(historyComponent());
		const input = screen.getByText(language.noHistory);
		expect(input).toBeInTheDocument();
	});

	it("render correct number of items in history", () => {
		render(historyComponent({ history }));
		const historyItems = screen.getAllByTestId("translation-item");
		expect(historyItems.length).toBe(history.length);
	});

	it("render loading text when loading", () => {
		render(historyComponent({ loading: true }));
		const loadingDiv = screen.getByText(language.loading);
		expect(loadingDiv).toBeInTheDocument();
	});

	it("calls speakTTS function onClick 'translation-item'", () => {
		window.speechSynthesis = jest.fn();
		speechSynthesis.getVoices = jest.fn().mockReturnValue([{ lang: "en" }]);
		audio.speakTTS = jest.fn();

		render(historyComponent({ history }));

		const translationItemDivs = screen.getAllByTestId("translation-item");
		fireEvent.click(translationItemDivs[0]);

		expect(audio.speakTTS).toHaveBeenCalled();
		expect(audio.speakTTS).toHaveBeenCalledWith(undefined, "en");
	});
});
