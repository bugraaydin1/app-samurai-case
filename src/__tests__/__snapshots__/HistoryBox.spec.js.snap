// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryBox renders correctly 1`] = `
<div
  className="history-wrapper"
  data-testid="history-wrapper"
>
  <h6
    className="sticky-title"
  >
    ðŸ‡¬ðŸ‡§
     
    Translation History
  </h6>
  <div>
    History not found
  </div>
</div>
`;
